### PR TITLE
Remove qajax, use axios consistently

### PR DESCRIPTION
**Pull request by milimetric**
So far, replaced Qajax, removed the dependency, checked that tests
aren't failing, and checked the interface superficially.  Next step is
to replace Q.Promise.

Looked into using fetch but a custom timeout and conversion of the
response to json is required, both of which axios does easily.  We could
implement these on top of fetch, but it seems cleaner this way.

The pull request that this commit will be part of works on #19 and #51

NOTE: axios is promising some good improvements in 0.19 but it's been
stuck in beta for a few months with no activity moving towards release.

### CODE REVIEW

**Comment by adrianmroz** [comment]
Could you elaborate on fetch vs axios? I don't have much experience with axios but I thought that because fetch is already in browser, why add another package. Links to worthy comparisons would be good also.

File: src/client/utils/ajax/ajax.ts [hunk 1]
```diff
@@ -16,9 +16,7 @@
  */
 
 import { ChainableExpression, Dataset, Environment, Executor, Expression, SplitExpression } from "plywood";
-import * as Qajax from "qajax";
-
-Qajax.defaults.timeout = 0; // We'll manage the timeout per request.
+import axios from "axios";
```
Comment (adrianmroz) [inline src/client/utils/ajax/ajax.ts]: reorder imports alphabetically, as linter suggests

File: src/client/utils/ajax/ajax.ts [hunk 2]
```diff
@@ -70,34 +60,30 @@ export class Ajax {
       if (Ajax.settingsVersionGetter) data.settingsVersion = Ajax.settingsVersionGetter();
     }
 
-    return Qajax({
+    return axios({
       method: options.method,
       url: options.url,
+      // Qajax's filterSuccess also allows 304s, axios by default does not
+      validateStatus: (s) => s >= 200 && s < 300 || s === 304,
```
Comment (adrianmroz) [inline src/client/utils/ajax/ajax.ts]: We don't use parens for single params in arrow functions - thus linter error.

**Comment by milimetric** [comment]
> Could you elaborate on fetch vs axios?

It's not too complicated of a decision, especially as far as Turnilo cares.  Basically axios just has all the features that I'd need to move from qajax without writing any custom code.  If we used fetch, I'd have to implement a timeout wrapper, so something like `fetchWithTimeout` that would call a `setTimeout` and throw an error.  There's also the matter of validating response statuses, which would have to be part of the same wrapper, and maybe a couple of other minor things as I go through the rest of the uses of Q.Promise.  Using fetch would also require a polyfill for IE (even the latest version).  There are a bunch of comparisons* but for turnilo it just comes down to: another library versus a polyfill and a bit of custom code.  I have no preference, and in both cases at least you'd have uniformity so it would be easy to move from one to the other.

My fault on the linting, didn't realize I had the config to lint locally.  I'll fix that up with the next push.

* example comparison: https://medium.com/@thejasonfile/fetch-vs-axios-js-for-making-http-requests-2b261cdd3af5

**Comment by adrianmroz** [comment]
Thanks, seems reasonable. If axios is simpler to move from q.ajax we should use it.

**Comment by milimetric** [comment]
Pushed the bigger part of this change, the refactor of q to native Promises.  So far it just builds, I haven't tested it at all, so I'll do one final commit to wrap this up once I verify.  Let me know if there are other automated tests or ways to verify, and whether you like the approach so far.

File: src/common/utils/ajax/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Compatibility with Q.defer or Promise.defer, now obsolete
+ */
+export class Deferred<T> {
```
Comment (adrianmroz) [inline src/common/utils/ajax/helpers.ts]: Cool that you move this into another file. Right now we used Q.defer as glorified callback and in short time we want to remove it. This way will be easier!

File: src/common/utils/ajax/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+  }
+}
+
+/**
+ * used everywhere that Q(null) was used
+ */
+export function noop(): Promise<any> {
```
Comment (adrianmroz) [inline src/common/utils/ajax/helpers.ts]: I would name it something like "resolveNull" or "resolveWithNull". It's not an "no operation", it's quite meaningful operation wich could cause side effects.

File: src/server/routes/mkurl/mkurl.ts [hunk 5]
```diff
@@ -81,8 +81,7 @@ router.post("/", (req: SwivRequest, res: Response) => {
       res.json({
         hash: `#${myDataCube.name}/${urlHashConverter.toHash(essence)}`
       });
-    })
-    .done();
```
Comment (adrianmroz) [inline src/server/routes/mkurl/mkurl.ts]: Why removing `done`? Is it safe?

File: src/server/utils/settings-manager/settings-manager.ts [hunk 6]
```diff
@@ -156,40 +155,43 @@ export class SettingsManager {
...
 
     var timeout = opts.timeout || this.initialLoadTimeout;
     if (timeout !== 0) {
-      currentWork = currentWork.timeout(timeout)
-        .catch(e => {
-          this.logger.error("Settings load timeout hit, continuing");
-        });
+      currentWork = Promise.race([
+        currentWork,
+        new Promise((resolve, reject) => {
```
Comment (adrianmroz) [inline src/server/utils/settings-manager/settings-manager.ts]: Could be usefull helper `timeout` which returns Promise that would resolve after x ms.

File: src/server/utils/settings-manager/settings-manager.ts [hunk 7]
```diff
@@ -156,40 +155,43 @@ export class SettingsManager {
     return this.timeMonitor.timekeeper;
   }
 
-  getSettings(opts: GetSettingsOptions = {}): Q.Promise<AppSettings> {
-    var currentWork = this.currentWork;
-
+  getSettings(opts: GetSettingsOptions = {}): Promise<AppSettings> {
     // Refresh all clusters
-    var currentWork = currentWork.then(() => {
+    var currentWork = this.currentWork.then(() => {
```
Comment (adrianmroz) [inline src/server/utils/settings-manager/settings-manager.ts]: Naming is confusing ...

File: src/server/utils/settings-store/settings-store.ts [hunk 8]
```diff
@@ -40,38 +40,36 @@ function readSettingsFactory(filepath: string, format: Format, inline = false) {
...
-        case "yaml":
-          return appSettingsToYAML(appSettings, false);
-        default:
-          throw new Error(`unsupported format '${format}'`);
-      }
-    })
-      .then(appSettingsYAML => {
-        return fs.writeFile(filepath, appSettingsYAML);
-      });
+    return noop()
```
Comment (adrianmroz) [inline src/server/utils/settings-store/settings-store.ts]: If I understand correctly, some branches in switch returns Promises and some not? I would prefer to call switch outside Promise and inside cases return Promise (like `Promise.resolve(JSON.stringify(...`). We would get rid of this one `then` and `noop`.

File: src/server/utils/settings-store/settings-store.ts [hunk 9]
```diff
@@ -92,13 +90,13 @@ export class SettingsStore {
...
-        .then(stateData => AppSettings.fromJS(JSON.parse(stateData), { visualizations: MANIFESTS }))
-      );
+      return stateStore.readState()
+        .then(stateData => AppSettings.fromJS(JSON.parse(stateData), { visualizations: MANIFESTS }));
     };
 
     settingsStore.writeSettings = (appSettings: AppSettings) => {
-      return Q.fcall(() => JSON.stringify(appSettings))
+      return noop()
+        .then(() => JSON.stringify(appSettings))
```
Comment (adrianmroz) [inline src/server/utils/settings-store/settings-store.ts]: Same here, `Promise.resolve(JSON.stringify...`

**Comment by milimetric** [comment]
Thanks very much for the review, @adrianmroz, I haven't forgotten about this.  I'll address your changes probably Friday, and do a rebase to see if there's anything new I need to patch up.

**Comment by adrianmroz** [comment]
Cool, keep closer look on #269.

File: package.json [hunk 10]
```diff
@@ -111,7 +110,6 @@
     "@types/node": "8.5.2",
     "@types/nopt": "3.0.29",
     "@types/q": "1.0.6",
-    "@types/qajax": "0.0.29",
```
Comment (l2dy) [inline package.json]: Fixed in #343 except for some dependency cleanup?

Comment (adrianmroz) [reply package.json]: Good catch!

**Comment by adrianmroz** [comment]
Closed with #343

**Comment by milimetric** [comment]
@adrianmroz I got back to this late, my bad.  Would you still like to see any of these changes or shall I try something else?  I am happy to finish this if you'd still like to merge it.  Otherwise maybe I'll take something smaller so I have a higher chance of getting it done.

**Comment by adrianmroz** [comment]
Most of your code was included in some way or another, so no worries! :) You can look around for issues with "help wanted" label.
